source/client/client_worker_impl.cc

```diff
@@ -22,8 +22,11 @@ void ClientWorkerImpl::simpleWarmup() {
   if (prefetch_connections_) {
     benchmark_client_->prefetchPoolConnections();
   }
-  benchmark_client_->tryStartOne([this] { dispatcher_->exit(); });
-  dispatcher_->run(Envoy::Event::Dispatcher::RunType::Block);
+  if (benchmark_client_->tryStartOne([this] { dispatcher_->exit(); })) {
```

Comment:
So I am new to this codebase so bare with me.  The name "tryStartOne" is a bit confusing.  The class is BenchmarkClient, but "One" here would seem to imply that you are starting a Client, but given that the function is not static or a factory that doesn't make sense.  How about tryStartOneRequest ?

The class BenchmarkClientHttpImpl has a method onComplete() which is part of StreamDecoderCompletionCallback.   AFAICT this is a per-request callback.   Therefore this class seems to be mixing "Client" which handles multiple requests and "Request" methods.  This only make sense if this client can only have a single outstanding request, but then the method would be tryStart rather than tryStartOneRequest and there would have to be many clients.

I would expect that there would be an object to manage the outstanding request and that object would have the onComplete handler which would then call back to the BenchmarkClientHttpImpl handler with a pointer to itself so that the BenchmarkClientHttpImpl could determine which request was failing/succeeding.  Possibly Nighthawk doesn't care specifically which request completes?

Reply:
Do the requests by necessity complete in order?  Does a Client manage a single connection or multiple connections?  HTTP/2 requests need not complete in-order so then the question remains how one can determine latency without knowing which request completed?

Reply:
Latency is captured at the stream level here:
https://github.com/envoyproxy/nighthawk/blob/master/source/client/stream_decoder.cc#L36

And at the sequencer level here (which contains queue/connect time): https://github.com/envoyproxy/nighthawk/blob/master/source/common/sequencer_impl.cc#L113

Reply:
Got it.  Thanx for the pointers!